ports/nrf/common-hal/digitalio/DigitalInOut.c

```diff
@@ -47,66 +50,56 @@ void common_hal_digitalio_digitalinout_deinit(digitalio_digitalinout_obj_t *self
     if (common_hal_digitalio_digitalinout_deinited(self)) {
         return;
     }
+    nrf_gpio_cfg_default(self->pin->number);
 
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-    nrf_gpio_cfg_default(pin);
-
+    reset_pin_number(self->pin->number);
     self->pin = mp_const_none;
 }
 
 void common_hal_digitalio_digitalinout_switch_to_input(
         digitalio_digitalinout_obj_t *self, digitalio_pull_t pull) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
-    nrf_gpio_cfg_input(pin, NRF_GPIO_PIN_NOPULL);
-
+    self->output = false;
+    nrf_gpio_cfg_input(self->pin->number, NRF_GPIO_PIN_NOPULL);
     common_hal_digitalio_digitalinout_set_pull(self, pull);
 }
 
 void common_hal_digitalio_digitalinout_switch_to_output(
         digitalio_digitalinout_obj_t *self, bool value,
         digitalio_drive_mode_t drive_mode) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
+    self->output = true;
     self->open_drain = (drive_mode == DRIVE_MODE_OPEN_DRAIN);
 
-    nrf_gpio_cfg_input(pin, NRF_GPIO_PIN_NOPULL);
+    nrf_gpio_cfg_input(self->pin->number, NRF_GPIO_PIN_NOPULL);
 
     common_hal_digitalio_digitalinout_set_value(self, value);
 }
 
 digitalio_direction_t common_hal_digitalio_digitalinout_get_direction(
         digitalio_digitalinout_obj_t *self) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
-    return (nrf_gpio_pin_dir_get(pin) == NRF_GPIO_PIN_DIR_OUTPUT) ? DIRECTION_OUTPUT : DIRECTION_INPUT;
+    return self->output ? DIRECTION_OUTPUT : DIRECTION_INPUT;
 }
 
 void common_hal_digitalio_digitalinout_set_value(
         digitalio_digitalinout_obj_t *self, bool value) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
     if (value && self->open_drain) {
-        nrf_gpio_pin_dir_set(pin, NRF_GPIO_PIN_DIR_INPUT);
+        nrf_gpio_pin_dir_set(self->pin->number, NRF_GPIO_PIN_DIR_INPUT);
     } else {
-        nrf_gpio_pin_dir_set(pin, NRF_GPIO_PIN_DIR_OUTPUT);
-        nrf_gpio_pin_write(pin, value);
+        nrf_gpio_pin_dir_set(self->pin->number, NRF_GPIO_PIN_DIR_OUTPUT);
+        nrf_gpio_pin_write(self->pin->number, value);
     }
 }
 
 bool common_hal_digitalio_digitalinout_get_value(
         digitalio_digitalinout_obj_t *self) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-    const nrf_gpio_pin_dir_t dir = nrf_gpio_pin_dir_get(pin);
-
-    if (dir == NRF_GPIO_PIN_DIR_INPUT) {
-        if (self->open_drain)
+    if (nrf_gpio_pin_dir_get(self->pin->number) == NRF_GPIO_PIN_DIR_INPUT) {
```

Comment:
To be consistent this should check self->output

Reply:
In this case we can recover the value from the pin state. In the previous case you can't tell the direction from the pin state. But I'll rework this as mentioned, I think in another PR.